src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -407,5 +406,176 @@ public static void SetLastWriteTime(string fullPath, DateTimeOffset time, bool a
+        }
+
+        internal static string? GetLinkTarget(string linkPath, bool isDirectory)
+            => GetImmediateLinkTarget(linkPath, isDirectory, throwOnNotFound: false, normalize: false);
+
+        /// <summary>
+        /// Gets reparse point information associated to <paramref name="linkPath"/>.
+        /// </summary>
+        /// <returns>The immediate link target, absolute or relative or null if the file is not a supported link.</returns>
+        internal static unsafe string? GetImmediateLinkTarget(string linkPath, bool isDirectory, bool throwOnNotFound, bool normalize)
+        {
+            using SafeFileHandle handle = OpenSafeFileHandle(linkPath,
+                    Interop.Kernel32.FileOperations.FILE_FLAG_BACKUP_SEMANTICS |
+                    Interop.Kernel32.FileOperations.FILE_FLAG_OPEN_REPARSE_POINT);
+
+            if (handle.IsInvalid)
+            {
+                int error = Marshal.GetLastWin32Error();
+                switch (error)
+                {
+                    case Interop.Errors.ERROR_FILE_NOT_FOUND:
+                    case Interop.Errors.ERROR_PATH_NOT_FOUND:
+                        if (throwOnNotFound)
+                        {
+                            throw Win32Marshal.GetExceptionForWin32Error(
+                                // File not found doesn't make much sense coming from a directory.
+                                isDirectory ? Interop.Errors.ERROR_PATH_NOT_FOUND : error, linkPath);
+                        }
+                        return null;
+                    default:
+                        throw Win32Marshal.GetExceptionForWin32Error(error, linkPath);
+                }
+            }
+
+            byte[]? buffer = null;
+            try
+            {
+                buffer = ArrayPool<byte>.Shared.Rent(Interop.Kernel32.MAXIMUM_REPARSE_DATA_BUFFER_SIZE);
+                bool success = Interop.Kernel32.DeviceIoControl(
+                    handle,
+                    dwIoControlCode: Interop.Kernel32.FSCTL_GET_REPARSE_POINT,
+                    lpInBuffer: IntPtr.Zero,
+                    nInBufferSize: 0,
+                    lpOutBuffer: buffer,
+                    nOutBufferSize: Interop.Kernel32.MAXIMUM_REPARSE_DATA_BUFFER_SIZE,
+                    out _,
+                    IntPtr.Zero);
+
+                if (!success)
+                {
+                    int error = Marshal.GetLastWin32Error();
+                    // The file or directory is not a reparse point.
+                    if (error == Interop.Errors.ERROR_NOT_A_REPARSE_POINT)
+                    {
+                        return null;
+                    }
+
+                    throw Win32Marshal.GetExceptionForWin32Error(error, linkPath);
+                }
+
+                ReadOnlySpan<byte> bufferSpan = new(buffer);
+                ref readonly Interop.Kernel32.REPARSE_DATA_BUFFER rdb = ref MemoryMarshal.AsRef<Interop.Kernel32.REPARSE_DATA_BUFFER>(bufferSpan);
```

Comment:
@ericstj I need some help figuring out how to fix the build failure coming from `Microsoft.IO.Redist`.

The type `MemoryMarshal` is detected in that project, but the `AsRef` method is not found. Should I:

- Include in the csprofj the file `src\libraries\System.Private.CoreLib\src\System\Runtime\InteropServices\MemoryMarshal.cs`?
- Add a reference to the `System.Runtime.InteropServices` assembly?
- Exclude these new APIs completely from Microsoft.IO.Redist? If yes, how do I do this?
- Something else? I couldn't find any references to `MemoryMarshal` inside that project, so I don't know what has been done in the past.

Note that this is the only issue left for me to address (locally).

Reply:
Option #1 seems quite problematic because importing the `MemoryMarshal.cs` file adds a lot of new errors about additional dependencies. I thought it could be an option because I noticed we are already importing another S.R.InteropServices file into M.IO.Redist:
https://github.com/dotnet/runtime/blob/44f050acc0f7791d6cd7ac772945444912bcf299/src/libraries/Microsoft.IO.Redist/src/Microsoft.IO.Redist.csproj#L193-L194

Reply:
@Jozkee and I talked and we are going to try to refactor this to use APIs available in net48.

Reply:
Should be fixed by https://github.com/dotnet/runtime/pull/54253/commits/24f8bfb88339699a34cd52cdff79f4e7f76b6885.